### ISSUE TITLE
[Feature request] "Circular" LinkedList

### ISSUE THREAD

**Issue by cyrilchapon** [Yomguithereal/mnemonist#228]
Hi ! How about a "Circular" kind of LinkedList ? Would be a good addition to that great library.

My personal use-case : building 2 apps : "secret-Santa" like and "killer party" like; which require a basic algorithm of "assigning targets" in a circular way

**Comment by mark-addams001881** [comment]
Hello, we tried to solve the issue.

This is what we did:

Added a new CircularLinkedList class that extends the existing LinkedList class. The new class overrides the push, unshift, and shift methods to maintain the circular structure. Also added a new rotate method to move elements around the circle.

You can review changes in this commit: https://github.com/mark-addams001881/Yomguithereal-mnemonist-228/commit/eb1de9b3748e3faafba842f12c30359db0986905.


> [!CAUTION]
> Disclaimer: The concept of solution was created by AI and you should never copy paste this code before you check the correctness of generated code. Solution might not be complete, you should use this code as an inspiration only.

---

Latta AI seeks to solve problems in open source projects as part of its mission to support developers around the world. Learn more about our mission at https://latta.ai/ourmission . If you no longer want Latta AI to attempt solving issues on your repository, you can block this account.

**Comment by Yomguithereal** [comment]
Hello @cyrilchapon. First of all, can you tell me whether a [`CircularBuffer`](https://yomguithereal.github.io/mnemonist/circular-buffer) would do the job here? Else, I see your point but I am not sure a new structure would add much vs. adding "circular-oriented" methods on the linked list itself, such as iterations starting from a specific node and such.

**Comment by cyrilchapon** [comment]
Hey @Yomguithereal ;
This might be a little naive but this was my process of thought :

I have x nodes
```
a            c

b            d
```

I need to "connect them" in some way, in a circular way.

```
a  -----  c
|              |
b  -----  d
```

Yeah basically
- Looping "from any element" for example, might be other use cases
- Ensuring like consistency of the loop (where it loops)

-------

To be perfectly honest, LinkedList logic was my first mental approach, but after posting I continued digging and researching (I'm not Math oriented so I'm struggling here); and found out some graph theory could be well suited for that. Like a "Directed Graph" with some "matching" algorithms (perfect matching, in my case).

**Comment by Yomguithereal** [comment]
Yes it looks like you either need something very specific, or some kind of graph data structure (like this one maybe: https://graphology.github.io/). It looks outside the scope of what I usually include in this library sorry :'(
